### PR TITLE
Treat hover events as normal pointer events, and bring them back to Listener

### LINKED ISSUE #33312: Add a hover GestureRecognizer
## Use case
The hover gesture is an important one on devices with mouse support, particularly when  designing flutter for web.  A HoverGestureRecognizer would make it easier, for example, to add a recognizer to a TextSpan.

## Proposal
Create a GestureRecognizer HoverGestureRecognizer, similar to TapGestureRecognizer.

### LINKED ISSUE #45910: Accessibility hover events should trigger onHover in MouseRegion.
Flutter doesn't respond to hover events in a11y contexts: i.e. single-tap vs double-tap in TalkBack.

We should connect a11y hover events through to simulate hover in MouseRegions.

### PR DESCRIPTION
**Pull request by dkwingsmt**
## Description

This PR changes the handling of hover events. Now they are dispatched the same way as regular events, going through regular hit tests, and are dispatched to `RenderObject.handleEvent` of all objects.

This PR also removes the long-deprecated `Listener.onPointerEnter` and `Listener.onPointerExit`, but un-deprecates `Listener.onPointerHover`, which is now the better way to receive hover events than `MouseRegion`.

### Current mechanism of hover events

For a long time Hover events have been dispatched only to `MouseTrackerAnnotation` render objects. This was designed because hover events are frequently triggered but would have a noticeable impact on performance if it went through the regular hit tests, back when hit tests were slow.

The performance is no longer an issue thanks to https://github.com/flutter/flutter/pull/59883, and now that mouse tracker already uses the regular hit test as well, we can dispatch hover events to all objects using the existing result.

There is one catch: events are dispatched after a transformation to the local coordinate space, regardless of whether the local information is used. This might cause a noticeable performance loss due to the frequency of hover events. This problem will be solved by https://github.com/flutter/flutter/pull/63813, which makes the transformation for local information lazy.

### Hover events from all devices are dispatched to all objects 

All devices, instead of only mice, can initiate valid hover events, which will be received by all objects instead of only `MouseTrackerAnnotation` objects. 

Allowing all objects to receive hover events will probably fix https://github.com/flutter/flutter/issues/45910, which demands to detect single taps in a11y mode, which are dispatched by the platform as hover events and have been ignored by gesture binding. (It's _not_ the first time we've met this problem: https://github.com/flutter/flutter/pull/46124 extended platform views with `MouseTrackingAnnotation` so that they can receive mouse events in a11y mode. It was good until recently an internal ticket (b/163613117) wanted widgets other than platform views to detect a11y single taps too.)

Allowing hover events from all devices will probably fix https://github.com/flutter/flutter/issues/33312, which demands gesture recognizers to be able to receive hover events. (Although no one has yet to bring up valid use cases for this feature.)

### Hover no longer part of MouseTrackerAnnotation

Since hover events are considered a regular event instead of "mouse tracked" event, `onHover` has been removed from `MouseTrackerAnnotation`. As a result, `MouseTrackerAnnotation` and `MouseTracker` only deals with logics that require keeping track of "what was hovered in the last frame" and computing the changes, i.e. entering, exiting, and cursors - a much clearer definition.

The `MouseRegion` still has `onHover` callback, which is now dispatched from `handleEvent`. Although we can remove this, I see no necessity for now.

### Listener

The `Listener` removed methods `onPointerEnter` and `onPointerExit` officially, a long overdue deprecation. However, the `onPointerHover` is un-deprecated, and is now a preferred way of detecting hover events over `MouseRegion.onHover`.

~~`Listener` no longer wraps `MouseRegion`. This results in the change to `print_user_created_ancestor_expectation.txt`.~~ This is moved to a separate PR due to breakage to g3.

### Order of hover events

The order between hover events is changed. They used to be in reverse visual order, but now in visual order, the same as other pointer events.

Hover events also start to respect `HitTestBehavior` when applicable; except for `MouseRegion`'s, which still respect `opaque` (we might change this in a future PR).

The order between different types of events are not changed. Hover events are still dispatched after exits and enters.

## Related Issues

- Fixes https://github.com/flutter/flutter/issues/45910
- Fixes https://github.com/flutter/flutter/issues/33312
- https://github.com/flutter/flutter/pull/63813, which fixes a performance issue brought by this PR.

## Tests

I added the following tests:

*Replace this with a list of the tests that you added as part of this PR. A change in behavior with no test covering it
will likely get reverted accidentally sooner or later. PRs must include tests for all changed/updated/fixed behaviors. See [Test Coverage].*

## Checklist

Before you create this PR, confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [ ] No, no existing tests failed, so this is *not* a breaking change.
- [x] Yes, this is a breaking change. *If not, delete the remainder of this section.*
   - [ ] I wrote a design doc: https://flutter.dev/go/template *Replace this with a link to your design doc's short link*
   - [ ] I got input from the developer relations team, specifically from: *Replace with the names of who gave advice*
   - [ ] I wrote a migration guide: https://flutter.dev/go/breaking-changes-template *Replace this with a link to a pull request that adds the migration guide to https://flutter.dev/docs/release/breaking-changes*

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

File: packages/flutter/lib/src/widgets/basic.dart [hunk 1]
```diff
@@ -6074,13 +5998,22 @@ class MouseRegion extends StatefulWidget {
   ///    internally implemented.
   final PointerEnterEventListener onEnter;
 
-  /// Triggered when a mouse pointer has moved onto or within the widget without
+  /// Triggered when a pointer changes into a position within this widget without
```
Comment (gspencergoog) [inline packages/flutter/lib/src/widgets/basic.dart:6001]:
```suggestion
  /// Triggered when a pointer moves to a position within this widget without
```

File: packages/flutter/lib/src/widgets/basic.dart [hunk 2]
```diff
@@ -6074,13 +5998,22 @@ class MouseRegion extends StatefulWidget {
...
   /// buttons pressed.
   ///
-  /// This callback is not triggered by the movement of an annotation.
+  /// Usually this is only fired for pointers which report their location when
+  /// not down (e.g. mouse pointers). Certain devices also fires this event on
```
Comment (gspencergoog) [inline packages/flutter/lib/src/widgets/basic.dart:6005]:
```suggestion
  /// not down (e.g. mouse pointers). Certain devices also fire this event on
```

**Review by gspencergoog — CHANGES_REQUESTED** [review]

**Comment by dkwingsmt** [comment]
@gspencergoog Thanks for the suggestions, I've fixed them

**Review by gspencergoog — APPROVED** [review]
32384589-a60f0e74-c078-11e7-9bc1-e5b5287aea9d

**Comment by dkwingsmt** [comment]
Update: This PR is blocked by https://github.com/flutter/flutter/pull/63813, which is further blocked by another issue.

**Comment by tvolkert** [comment]
> The performance is no longer an issue thanks to #59883

But this still causes us to hit test the whole tree on every mouse move event, right?

**Comment by tvolkert** [comment]
This PR is listed as a breaking change, but there's no link to a design doc or migration guide.

**Comment by dkwingsmt** [comment]
@tvolkert 

> But this still causes us to hit test the whole tree on every mouse move event, right?

Yes.

> This PR is listed as a breaking change, but there's no link to a design doc or migration guide.

That might have been a misjudgement according to the standard in the wiki. It didn't cause breakages in internal or flutter/tests.

**Pull request merged**